### ISSUE TITLE
Grayed / greyed out / inactive chat messages

### ISSUE THREAD

**Issue by taoeffect** [okTurtles/group-income#2061]
### Problem

![Screenshot 2024-06-11 at 12 29 59 PM](https://github.com/okTurtles/group-income/assets/138706/aeb656b4-6e59-432e-b718-b8f5429e6dc2)

With the latest deployment, this is still a problem. For whatever reason the message appears grayed out after being sent.

NOTE: simply switching to the dashboard and back to the chat shows the message as active.

### Solution

It might be that some `.$forceUpdate`s that were removed in previous PRs might've actually been necessary. Although it could also be something else, please see comments below.

**Comment by Silver-IT** [comment]
@taoeffect, Do you think it happened only in PWA?

**Comment by Silver-IT** [comment]
To be honest, I don't think [$forceUpdate](https://vuejs.org/api/component-instance.html#forceupdate) function is solution for this issue, because the `this.messages` is not non-reactive component state. It is the `computed value` of the component which uses the reactive state which are declared in component `data` option.

**Comment by taoeffect** [comment]
@Silver-IT 

> @taoeffect, Do you think it happened only in PWA?

No, this happened on Desktop Brave.

> To be honest, I don't think [$forceUpdate](https://vuejs.org/api/component-instance.html#forceupdate) function is solution for this issue

Then why does switching the page to the dashboard and back fix it?

**Comment by taoeffect** [comment]
More info on this. Here it's happening on the testing site right now with v0.5.3 in the `#general` chatroom. Notice that each time I send a message the previous message that was greyed out becomes ungrayed out, but the latest message remains inactive:

https://github.com/okTurtles/group-income/assets/138706/1164dfd1-3a76-45e9-8db1-1636afe0c987

Meanwhile, on the server, we have errors that look like this appearing:

```
[19:58:46.446] ERROR (74): [chelonia] ERROR in handleEvent: Message z9brRu3VSrqvvKWZjzZCYzcEkiWue2qSngDMHh9zsbs1LZKL7SNV with height 23 in contract z9brRu3VQnrQ3EH5py4oFGPJHAmBk3o1S6UskgSqEFiXvekuHoLV has already been processed. Current height: 23. {"name":"ChelErrorAlreadyProcessed"}                                
[19:58:46.446] ERROR (74): ChelErrorAlreadyProcessed                                                                                                                                                                                                                                                                         
    err: {                                                                                                                                                                                                                                                                                                                   
      "type": "",                                                                                                                                                                                                                                                                                                            
      "message": "Message z9brRu3VSrqvvKWZjzZCYzcEkiWue2qSngDMHh9zsbs1LZKL7SNV with height 23 in contract z9brRu3VQnrQ3EH5py4oFGPJHAmBk3o1S6UskgSqEFiXvekuHoLV has already been processed. Current height: 23.",                                                                                                             
      "stack":                                                                                                                                                                                                                                                                                                               
          ChelErrorAlreadyProcessed: Message z9brRu3VSrqvvKWZjzZCYzcEkiWue2qSngDMHh9zsbs1LZKL7SNV with height 23 in contract z9brRu3VQnrQ3EH5py4oFGPJHAmBk3o1S6UskgSqEFiXvekuHoLV has already been processed. Current height: 23.                                                                                            
              at Object.call (/home/unpriv/gi/shared/domains/chelonia/internals.js:1827:13)                                                                                                                                                                                                                                  
              at Object.chelonia/private/in/handleEvent (/home/unpriv/gi/shared/domains/chelonia/internals.js:1706:56)                                                                                                                                                                                                       
              at sbp (/home/unpriv/gi/node_modules/@sbp/sbp/dist/main.cjs:50:32)                                                                                                                                                                                                                                             
              at invocation (/home/unpriv/gi/shared/domains/chelonia/internals.js:1118:16)                                                                                                                                                                                                                                   
              at Object.chelonia/private/invoke (/home/unpriv/gi/shared/domains/chelonia/internals.js:164:14)                                                                                                                                                                                                                
              at sbp (/home/unpriv/gi/node_modules/@sbp/sbp/dist/main.cjs:50:32)                                                                                                                                                                                                                                             
              at Object.okTurtles.eventQueue/queueEvent (/home/unpriv/gi/node_modules/@sbp/okturtles.eventqueue/dist/umd/index.cjs:79:60)                                                                                                                                                                                    
              at sbp (/home/unpriv/gi/node_modules/@sbp/sbp/dist/main.cjs:50:32)                                                                                                                                                                                                                                             
              at Object.chelonia/private/queueEvent (/home/unpriv/gi/shared/domains/chelonia/internals.js:170:15)                                                                                                                                                                                                            
              at sbp (/home/unpriv/gi/node_modules/@sbp/sbp/dist/main.cjs:50:32)                                                                                                                                                                                                                                             
              at Object.chelonia/private/in/enqueueHandleEvent (/home/unpriv/gi/shared/domains/chelonia/internals.js:1117:15)                                                                                                                                                                                                
              at sbp (/home/unpriv/gi/node_modules/@sbp/sbp/dist/main.cjs:50:32)                                                                                                                                                                                                                                             
              at Object.backend/server/handleEntry (/home/unpriv/gi/backend/server.js:140:14)                                                                                                                                                                                                                                
              at sbp (/home/unpriv/gi/node_modules/@sbp/sbp/dist/main.cjs:50:32)                                                                                                                                                                                                                                             
              at /home/unpriv/gi/backend/routes.js:71:16                                                                                                                                                                                                                                                                     
              at exports.Manager.execute (/home/unpriv/gi/node_modules/@hapi/hapi/lib/toolkit.js:57:29)                                                                                                                                                                                                                      
              at Object.internals.handler (/home/unpriv/gi/node_modules/@hapi/hapi/lib/handler.js:46:48)                                                                                                                                                                                                                     
              at exports.execute (/home/unpriv/gi/node_modules/@hapi/hapi/lib/handler.js:31:36)                                                                                                                                                                                                                              
              at Request._lifecycle (/home/unpriv/gi/node_modules/@hapi/hapi/lib/request.js:372:68)                                                                                                                                                                                                                          
              at processTicksAndRejections (node:internal/process/task_queues:95:5)                                                                                                                                                                                                                                          
              at Request._execute (/home/unpriv/gi/node_modules/@hapi/hapi/lib/request.js:280:9)                                                                                                                                                                                                                             
      "name": "ChelErrorAlreadyProcessed"                                                                                                                                                                                                                                                                                    
    }                                                   
[19:58:46.447] ERROR (74): Debug: handler, error                                                                                                                                                                                                                                                                             
    Error: Message z9brRu3VSrqvvKWZjzZCYzcEkiWue2qSngDMHh9zsbs1LZKL7SNV with height 23 in contract z9brRu3VQnrQ3EH5py4oFGPJHAmBk3o1S6UskgSqEFiXvekuHoLV has already been processed. Current height: 23.
    at /home/unpriv/gi/backend/routes.js:101:24                                                                                                               
    at processTicksAndRejections (node:internal/process/task_queues:95:5)                                                                                     
    at exports.Manager.execute (/home/unpriv/gi/node_modules/@hapi/hapi/lib/toolkit.js:60:28)                                                                                                                                                                                                                                
    at Object.internals.handler (/home/unpriv/gi/node_modules/@hapi/hapi/lib/handler.js:46:20)                                                                                                                                                                                                                               
    at exports.execute (/home/unpriv/gi/node_modules/@hapi/hapi/lib/handler.js:31:20)                                                                                                                                                                                                                                        
    at Request._lifecycle (/home/unpriv/gi/node_modules/@hapi/hapi/lib/request.js:372:32)                                                                                                                                                                                                                                    
    at Request._execute (/home/unpriv/gi/node_modules/@hapi/hapi/lib/request.js:280:9) 
[19:58:52.248] ERROR (74): [chelonia] ERROR in handleEvent: Message z9brRu3VPcKoJHthcmewgh3eopbdbtUikNqgRno3XZ1ERDjm33Fk with height 24 in contract z9brRu3VQnrQ3EH5py4oFGPJHAmBk3o1S6UskgSqEFiXvekuHoLV has already been processed. Current height: 24. {"name":"ChelErrorAlreadyProcessed"}
[19:58:52.248] ERROR (74): ChelErrorAlreadyProcessed                                                                                                            
    err: {
      "type": "",
      "message": "Message z9brRu3VPcKoJHthcmewgh3eopbdbtUikNqgRno3XZ1ERDjm33Fk with height 24 in contract z9brRu3VQnrQ3EH5py4oFGPJHAmBk3o1S6UskgSqEFiXvekuHoLV has already been processed. Current height: 24.",
      "stack":
          ChelErrorAlreadyProcessed: Message z9brRu3VPcKoJHthcmewgh3eopbdbtUikNqgRno3XZ1ERDjm33Fk with height 24 in contract z9brRu3VQnrQ3EH5py4oFGPJHAmBk3o1S6UskgSqEFiXvekuHoLV has already been processed. Current height: 24.
              at Object.call (/home/unpriv/gi/shared/domains/chelonia/internals.js:1827:13)                                                                     
              at Object.chelonia/private/in/handleEvent (/home/unpriv/gi/shared/domains/chelonia/internals.js:1706:56)                                          
              at sbp (/home/unpriv/gi/node_modules/@sbp/sbp/dist/main.cjs:50:32)                                                                                
              at invocation (/home/unpriv/gi/shared/domains/chelonia/internals.js:1118:16)                                                                      
              at Object.chelonia/private/invoke (/home/unpriv/gi/shared/domains/chelonia/internals.js:164:14)                                                   
              at sbp (/home/unpriv/gi/node_modules/@sbp/sbp/dist/main.cjs:50:32)                                                                                
              at Object.okTurtles.eventQueue/queueEvent (/home/unpriv/gi/node_modules/@sbp/okturtles.eventqueue/dist/umd/index.cjs:79:60)                       
              at sbp (/home/unpriv/gi/node_modules/@sbp/sbp/dist/main.cjs:50:32)                                                                                
              at Object.chelonia/private/queueEvent (/home/unpriv/gi/shared/domains/chelonia/internals.js:170:15)                                               
              at sbp (/home/unpriv/gi/node_modules/@sbp/sbp/dist/main.cjs:50:32)                                                                                
              at Object.chelonia/private/in/enqueueHandleEvent (/home/unpriv/gi/shared/domains/chelonia/internals.js:1117:15)                                   
              at sbp (/home/unpriv/gi/node_modules/@sbp/sbp/dist/main.cjs:50:32)                                                                                
              at Object.backend/server/handleEntry (/home/unpriv/gi/backend/server.js:140:14)                                                                   
              at sbp (/home/unpriv/gi/node_modules/@sbp/sbp/dist/main.cjs:50:32)                                                                                
              at /home/unpriv/gi/backend/routes.js:71:16                                                                                                        
              at exports.Manager.execute (/home/unpriv/gi/node_modules/@hapi/hapi/lib/toolkit.js:57:29)                                                         
              at Object.internals.handler (/home/unpriv/gi/node_modules/@hapi/hapi/lib/handler.js:46:48)                                                        
              at exports.execute (/home/unpriv/gi/node_modules/@hapi/hapi/lib/handler.js:31:36)                                                                 
              at Request._lifecycle (/home/unpriv/gi/node_modules/@hapi/hapi/lib/request.js:372:68)                                                             
              at processTicksAndRejections (node:internal/process/task_queues:95:5)                                                                             
              at Request._execute (/home/unpriv/gi/node_modules/@hapi/hapi/lib/request.js:280:9)                                                                
      "name": "ChelErrorAlreadyProcessed"                                                                                                                       
    }

**Comment by corrideat** [comment]
Although I'm not sure yet why a sending new message is needed to unmark a previous message as pending, this issue would seem to indicate **some sort of issue with receiving events over the web socket**. This could be: (1) in ChatMain (unlikely, because messages are failing to send), (2) in Chelonia (for example, the message is being ignored or processing fails; possible but not as likely since the logs don't show this) or (3) in the websocket itself (for example, the server is not sending the event).

Note that messages fail to send twice but then appear to be sent, going by the logs. This is consistent with the mechanism in `publishEvent` to handle these situations, which results in calling `chelonia/contract/sync` to retrieve the correct height. Although this causes the message to be sent, calling `chelonia/contract/sync` shouldn't be needed, and certainly not calling `chelonia/contract/sync` repeatedly.

Further investigation is still needed to determine the cause. I was able to reproduce earlier but not today.

**Comment by taoeffect** [comment]
Possibly fixed by the fix for #2128. If anyone runs into this again please let us know!
